More refactoring of routing components

imalberto opened the pull request:
CHANGELOG:
- Now using `express-annotations` to augment route configuration with meta data
- Now using `express-map` to generate route config format used throughout mojito
- Now using `express-state` `serialize` method to replace `JSON.stringify` for serializing objects to client runtime
- Refactored set of components to deal with revised route configuration format generated by `express-map`
- Updated `dispatcher.js`, `router.js` to take advantage of `express-map` and `express-annotations`
- Updated `ac.url` addon and `route-maker` to leverage `express-map`

TODO:
- [x] fix unit tests
- [x] fix func tests
- [x] update examples
- [x] publish most current version of `express-annotations` to npm
- [x] publish most current version of `express-map` to npm
- [x] merge with upstream (next branch)

imalberto committed:
First pass at integrating express-map and express-routes.
Other changes include:
- doc api updates
- multiple minor code refactoring

imalberto committed:
First pass at integrating express-map and express-routes.
Other changes include:
- doc api updates
- multiple minor code refactoring

imalberto committed:
Now using `express-map` to created named route configuration.

imalberto committed:
Now using `express-map` to created named route configuration.

imalberto committed:
Simplified route configuration serialization between add ons and client runtime.

imalberto committed:
Simplified route configuration serialization between add ons and client runtime.

imalberto committed:
- Added RegExp to whitelist for cleanse
- No more converting route configuration to "old format"
- Now using express-state/lib/serialize

imalberto committed:
- Added RegExp to whitelist for cleanse
- No more converting route configuration to "old format"
- Now using express-state/lib/serialize

imalberto committed:
- Updated route lookup based on express-map pr#4
- No need to explicitly copy over the `annotations` property now that routes configuration is read from express-map#getRouteMap()

imalberto committed:
- Updated route lookup based on express-map pr#4
- No need to explicitly copy over the `annotations` property now that routes configuration is read from express-map#getRouteMap()

imalberto committed:
Removed unused code.

imalberto committed:
Removed unused code.

imalberto committed:
Deleted unused code.

imalberto committed:
Deleted unused code.

caridy commented on lib/mojito.js:
```diff
@@ -69,6 +71,11 @@ Mojito extension for Express.
 **/
 function Mojito(app) {
 
+    // FIXME: This might not be required since app.js would have already
```
I don't think we should worry about this. If they try to extend it manually, let the lib to handle that :)

caridy commented:
This is looking good!

imalberto committed:
Updated from PR feedback

imalberto committed:
Updated from PR feedback

ericf commented on lib/router.js:
```diff
@@ -161,26 +161,17 @@ module.exports = {
+                    debug('[%s %s] installing handler', route.call, route.path);
+                    verb = verb.toLowerCase();
+
+                    app[verb](route.path, dispatcher.dispatch(route.call));
```
I think right after the route is created, we could add the `dispatch` annotation. `dispatcher.dispatch()` returns a route handler function with a `dispatch` property, we could capture this return value and add the annotation here. I think this will be much clearer than adding the annotation it inside the `dispatcher.js`.

``` javascript
var handler = dispatcher.dispatch(route.call);

app[verb](route.path, handler);
app.annotate(route.path, { dispatch: handler.dispatch });
```

Maybe there's some timing thing I'm missing, but this means you could remove the iteration over all the routes, the `init` state flag, etc from `dispatcher.js`.

imalberto replied:
There are 2 ways to register routes: using `app.VERB` and via `routes.json`. In this case, the annotations is Mojito specific, and there is no need to ask applications to attach some "annotations" that they are not going to use but is needed by the fw. By doing so, registering a route is same as any Express app:

`app.get('/foo', dispatcher('foo.index'));`

Whatever data the  `dispatcher` needs is kept private from the app.

ericf commented on lib/dispatcher.js:
```diff
@@ -65,32 +69,18 @@ module.exports = {
-                        mojito.routes[method] = cache;
-                    }
+            // Do this only once
+            if (!init) {
```
I think all of this can be removed but using the the approach [here in `router.js`](https://github.com/yahoo/mojito/pull/1233/files#r6080885).

imalberto replied:
Synced up with @ericf on this. Please see [above](https://github.com/yahoo/mojito/pull/1233/files#r6080885) in `router.js`.

ericf commented on lib/dispatcher.js:
```diff
@@ -160,12 +150,12 @@ module.exports = {
+        if (!CACHE.routes) {
+            // TODO: this could be problematic if the app adds additional
+            // routes at runtime.
+            CACHE.routes = app.getRouteMap();
```
I'd hope no one would run into this issue. But if we're worried about it, we could always override `app._router.route()` to blow alway the cache, this way when the next request comes in, the app's route map will be re-generated and cached.

imalberto replied:
@ericf Thanks for the suggestions. For this use case, it will be required for the app to do something like `dispatcher.resetRoutes()` as a prerequisite before modifying the routes configuration.

ericf commented:
@imalberto this is looking good!

Looking at `route-maker.common.js` I was hoping that more of the APIs from `express-annotations` and `express-map` could be used, then I noticed the `.common.js`. How much of Router Maker is used client-side?

I think it would be good to talk about the annotation strategy, i.e. using simple annotation values instead of objects, like `dispatch`, instead there could be separate annotations for `call`, `params`, `options`. Also, I was thinking that `express-map`'s `pathTo()` factory could be used to by the app to create a URL generator function that could be passed to Route Maker.

imalberto commented:
@ericf Re: your [previous](https://github.com/yahoo/mojito/pull/1233#issuecomment-23546468) comment, will look into how Mojito can leverage more of `express-map` and `express-annotations` in upcoming PR to address how the `pathTo` functionality can be exposed to handlebars helper (both client and server runtime).

imalberto committed:
- Added `resetRoutesConfig` for force Mojito to reload route configuration at runtime.
- Unit tests fixes

imalberto committed:
- Added `resetRoutesConfig` for force Mojito to reload route configuration at runtime.
- Unit tests fixes

imalberto committed:
Added optional second argument to `dispatcher.dispatch` to pass any static route params.

imalberto committed:
Added optional second argument to `dispatcher.dispatch` to pass any static route params.

imalberto committed:
More tests fixes.

imalberto committed:
More tests fixes.

drewfish committed:
version of PR1236 (hotswap support) for "next" branch

Commit:
Merge pull request #1239 from lzhan/next

update app to meet next changes

imalberto committed:
- Revised proposal based on updated api from `express-map` and `express-annotations`
- Added sugar `app.mojito.get('foo', '/foo', 'foo.index');` to register routing with Mojito

imalberto committed:
- Revised proposal based on updated api from `express-map` and `express-annotations`
- Added sugar `app.mojito.get('foo', '/foo', 'foo.index');` to register routing with Mojito

imalberto committed:
Updated HISTORY_next.md with deprecated APIs

imalberto committed:
Updated HISTORY_next.md with deprecated APIs

imalberto committed:
- 2nd pass over the routing api
- Updated apidocs
- More unit tests fixes

imalberto committed:
- 2nd pass over the routing api
- Updated apidocs
- More unit tests fixes

caridy commented on lib/mojito.js:
```diff
@@ -234,8 +241,28 @@ Mojito.prototype._configureAppInstance = function (app, store, options) {
-        // @see lib/dispatcher.js
+    });
+
+    // Sugar example usage:
```
Can you mark this as experimental?

caridy replied:
or just remove it for now.

caridy commented on lib/mojito.js:
```diff
@@ -234,8 +241,28 @@ Mojito.prototype._configureAppInstance = function (app, store, options) {
+
+    // Sugar example usage:
+    //
+    //     app.mojito.get('foo', '/foo', 'foo.index', params);
```
I was thinking on something less sweet! something like:

```
app.mojito.get('foo', '/foo', middleware1, middleware2, dispatch('foo.index', { params }))
```

caridy commented on lib/router.js:
```diff
@@ -161,26 +163,18 @@ module.exports = {
+                    debug('[%s %s] installing handler', route.call, route.path);
+                    verb = verb.toLowerCase();
+
+                    app[verb](route.path, dispatcher.dispatch(route.call, route.params));
```
if this block is not going to use the sugar method, better to remove the sugar then until people start asking for it :)

imalberto replied:
Will remove the sugar method.

caridy commented on lib/app/autoload/util.common.js:
```diff
@@ -608,6 +612,45 @@ YUI.add('mojito-util', function(Y, NAME) {
+         * @param {String} key
+         * @return {Object}
+         */
+        decodeRouteName: function(key) {
```
why do we need this? this is a one way operation to generate a token, but the token will never be use to generate the original entries used to generate the token. remove.

caridy commented on lib/app/autoload/util.common.js:
```diff
@@ -125,6 +125,10 @@ YUI.add('mojito-util', function(Y, NAME) {
             // is found via different indexing models. If we got anything else
             // we can just return it.
 
+            if (Object.prototype.toString.apply(obj) === "[object RegExp]") {
```
what is this?

imalberto replied:
This is to allow the `ac.deploy` addon to allow `regex` to be serialized to the client using `express-state` as part of the cleanse process.

caridy replied:
ok, I never fully understood why the whole escaping of the serialized token. But now that we rely on `express-state`, I will argue that `express-state` is responsible for any escaping process needed to expose data into the client side as part of the html response. /cc @ericf

imalberto replied:
ok. if @ericf confirms, we can remove it.

caridy replied:
@drewfish @isao do you guys remember what was the problem with the whole encoding thing?

isao replied:
If memory serves cleanse/uncleanse was for protecting against XSS in an app's configs that were sent to client. IMO, it shouldn't apply to things the developer has defined. But perhaps there was a concern developers would put user input in the config object destined for the client at runtime.

caridy replied:
oh, right, I remember now. Well, at this point there is only few things that might be runtime data specific, so we can constrain that to those elements to avoid doing the full sanitizing process on all data we sent. Let's create a card for that @imalberto. Also, how expensive if this operation today?

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -313,29 +191,63 @@ YUI.add('mojito-route-maker', function (Y, NAME) {
-        @return {Object} the route configuration object
+        @return {Object} the route object or null
         **/
         find: function(uri, verb) {
```
please remove!

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -73,213 +73,90 @@ YUI.add('mojito-route-maker', function (Y, NAME) {
+            var parts = query.split('?'),
                 call = parts[0],
-                residual = {},
+                residual = { },
```
{}

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -73,213 +73,90 @@ YUI.add('mojito-route-maker', function (Y, NAME) {
-        @return {Object} route object matching the query, or null
+        @return {String} the generated path uri for the query, or null
         **/
         make: function(query, verb, params) {
```
is this api really changing or is this just a mismatch in the old doc?

caridy replied:
can u comment on this question?

imalberto replied:
`make()` returns a URI, so was probably a mismatch.

imalberto commented:
Also, after discussion with @caridy , will also be removing annotations and the `find()` method as part of `route-maker` and `ac.url` addon.

drewfish committed:
Merge pull request #1240 from drewfish/onestore

version of PR1236 (hotswap support) for "next" branch

imalberto committed:
- Removed all references to 'find()' methods
- Removed all references to 'annotations' - no need to serialize to clients
- Removed 'decodeRouteName()' since it is a one way operation
- Removed sugar methods for the router
-

imalberto committed:
- Removed all references to 'find()' methods
- Removed all references to 'annotations' - no need to serialize to clients
- Removed 'decodeRouteName()' since it is a one way operation
- Removed sugar methods for the router
-

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -42,21 +41,18 @@ YUI.add('mojito-route-maker', function (Y, NAME) {
     @param {String} uri
     **/
-    function doCallReplacement(route, uri) {
+    function resolveRouteParams(route, uri) {
```
do we really need this? I thought express-map will do all this heavy lifting.

imalberto replied:
Not on the client unfortunately. I mentioned that to @ericf and in the end, we have might a solution that works on both run time, but for now this is needed and is abstracted by `route-maker` with common affinity.

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -158,128 +92,68 @@ YUI.add('mojito-route-maker', function (Y, NAME) {
-            return match;
+        linkTo: function(name) {
+            // TODO How can we leverage express-map#pathTo ?
+            // return (CACHE.routes[name] && CACHE.routes[name].path) || null;
```
`require('express-map/lib/pathto')`?

imalberto replied:
Won't work on the client using this. Was going to implement this method as part of next PR to provide helpers for the built in templates. And it has to run on both run time.

caridy commented on lib/dispatcher.js:
```diff
@@ -16,9 +16,12 @@ Given a `command`, dispatches it to the Mojito engine.
     debug = require('debug')('mojito:dispatcher'),
-    OutputHandler = require('./output-handler.server');
+    OutputHandler = require('./output-handler.server'),
+    CACHE = { };
```
u keep using space in between empty `{}` lol.

caridy commented on lib/dispatcher.js:
```diff
@@ -27,6 +30,13 @@ var liburl = require('url'),
     /**
+    Provides a way to reset the internal routes configurations
+    **/
+    resetRoutesConfig: function () {
```
nice, I bet this is really useful for testing at least!

caridy commented on lib/dispatcher.js:
```diff
@@ -37,19 +47,24 @@ module.exports = {
-        // annotations, for now looking at the last one is just fine.
+        if (params && typeof params === 'string') {
+            routeParams = libqs.parse(params);
+        }
```
if it is not a string, but an object, what do u do with it?

imalberto replied:
Good catch. Should set `routeParams = params || {};` above.

caridy replied:
I literally made you to get in a trap! Now in line `93` you're augmenting the original `params` object per request, but `params` is actually an app level value, leaking data between requests :), solution will be described in lined `93`. jejeje

caridy commented on lib/router.js:
```diff
@@ -147,12 +146,15 @@ module.exports = {
             mojito = app.mojito, // the mojito instance
-            appRoot = mojito.options.root;
+            appRoot = mojito.options.root,
+            key = mojito.Y.mojito.util.encodeRouteName;
```
confusing, better to just call it encodeRouteName as a local variable, if that's the main objective here.

imalberto replied:
Will rename as suggested.

caridy commented on lib/app/autoload/util.common.js:
```diff
@@ -608,6 +612,21 @@ YUI.add('mojito-util', function(Y, NAME) {
+         * @param {String} call
+         * @return {String} key
+         */
+        encodeRouteName : function(verb, call) {
```
linting is needed here, space between func name and :

imalberto replied:
Agreed. Problem is most of those files have `sloppy: true`; At some point those should be removed in favor of lest lint config. The lack of space in this case is to be consistent in that file. Otherwise some function will have space, and others won't . They should all be cleaned up in one go, in a separate PR maybe.

imalberto committed:
Changes based on PR#1233 feedback.

imalberto committed:
Changes based on PR#1233 feedback.

caridy commented on lib/dispatcher.js:
```diff
@@ -101,7 +90,7 @@ module.exports = {
             command.context = req.context;
             command.params = {
-                route: req.params || {},
+                route: libutil.extend(routeParams, req.params) || {},
```
```
route: libutil.extend({}, routeParams, req.params),
```

will work better :), avoiding the leak.

imalberto committed:
Fix leaking parameters to caller!

imalberto committed:
Fix leaking parameters to caller!

caridy committed:
history entries for v0.7.4

imalberto committed:
Unit tests fixes.

imalberto committed:
Unit tests fixes.

caridy commented:
+1

imalberto committed:
- More functional and unit tests fixes.
- Deleted tests related to `route-maker#find()` and `ac.url#find()`
-

imalberto committed:
- More functional and unit tests fixes.
- Deleted tests related to `route-maker#find()` and `ac.url#find()`
-

imalberto committed:
Added express-state and express-map as dependencies

imalberto committed:
Added express-state and express-map as dependencies

imalberto committed:
Updated HISTORY_next with latest api revisions

imalberto committed:
Updated HISTORY_next with latest api revisions

imalberto committed:
Fix issue on startup where mojito.js trying to use appCofig.appPort before loading calling getAppConfig.

imalberto committed:
More func tests fixes.

imalberto committed:
Fixing runtime error is causing the entire suite to fail.

imalberto committed:
Merge downstream from remote/origin

imalberto commented:
# Unit Test Report Summary

Total Number of Executed Tests: 487
Total Number of Passed Tests: 475
Total Number of Failed Tests: 0
Total Number of Skipped Tests: 12
